air-script/tests/helpers/plonky3_periodic_columns.rs

```diff
@@ -0,0 +1,146 @@
+use p3_air::{Air, AirBuilder, AirBuilderWithPublicValues, BaseAir};
+use p3_field::Field;
+use p3_matrix::{
+    Matrix,
+    dense::{RowMajorMatrix, RowMajorMatrixView},
+    stack::VerticalPair,
+};
+
+pub trait BaseAirWithPeriodicColumns<F>: BaseAir<F> {
+    fn get_periodic_columns(&self) -> Vec<Vec<F>> {
+        vec![]
+    }
+}
+
+pub trait AirBuilderWithPeriodicColumns: AirBuilder {
+    type PeriodicColumnsVar: Field + Into<Self::Expr>;
+
+    fn periodic_columns(&self) -> Vec<Self::PeriodicColumnsVar> {
+        vec![]
+    }
+}
+
+pub(crate) fn check_constraints_with_periodic_columns<F, A>(
+    air: &A,
+    main: &RowMajorMatrix<F>,
+    public_values: &Vec<F>,
+) where
+    F: Field,
+    A: for<'a> Air<DebugConstraintBuilderWithPeriodicColumns<'a, F>>
+        + BaseAirWithPeriodicColumns<F>,
+{
+    let height = main.height();
+
+    (0..height).for_each(|i| {
+        let i_next = (i + 1) % height;
+
+        let local = main.row_slice(i).unwrap(); // i < height so unwrap should never fail.
+        let next = main.row_slice(i_next).unwrap(); // i_next < height so unwrap should never fail.
+        let main = VerticalPair::new(
+            RowMajorMatrixView::new_row(&*local),
+            RowMajorMatrixView::new_row(&*next),
+        );
+        let periodic_columns = air.get_periodic_columns();
+
+        let mut builder = DebugConstraintBuilderWithPeriodicColumns {
+            row_index: i,
+            main,
+            public_values,
+            is_first_row: F::from_bool(i == 0),
+            is_last_row: F::from_bool(i == height - 1),
+            is_transition: F::from_bool(i != height - 1),
+            periodic_columns,
+        };
+
+        air.eval(&mut builder);
+    });
+}
+
+/// A builder that runs constraint assertions during testing.
+///
+/// Used in conjunction with [`check_constraints`] to simulate
+/// an execution trace and verify that the AIR logic enforces all constraints.
+#[derive(Debug)]
+pub struct DebugConstraintBuilderWithPeriodicColumns<'a, F: Field> {
+    /// The index of the row currently being evaluated.
+    row_index: usize,
+    /// A view of the current and next row as a vertical pair.
+    main: VerticalPair<RowMajorMatrixView<'a, F>, RowMajorMatrixView<'a, F>>,
+    /// The public values provided for constraint validation (e.g. inputs or outputs).
+    public_values: &'a [F],
+    /// A flag indicating whether this is the first row.
+    is_first_row: F,
+    /// A flag indicating whether this is the last row.
+    is_last_row: F,
+    /// A flag indicating whether this is a transition row (not the last row).
+    is_transition: F,
+    /// The periodic columns provided for constraint validation.
+    periodic_columns: Vec<Vec<F>>,
+}
+
+impl<'a, F> AirBuilderWithPeriodicColumns for DebugConstraintBuilderWithPeriodicColumns<'a, F>
+where
+    F: Field + Into<Self::Expr>,
+{
+    type PeriodicColumnsVar = F;
+
+    fn periodic_columns(&self) -> Vec<Self::PeriodicColumnsVar> {
```

Comment:
While maybe not for this PR, I think we can represent `periodic_columns` in the `DebugConstraintBuilderWithPeriodicColumns` struct a bit differently
- Pad all periodic columns to the maximum length by periodic repetition
- Transpose the columns so that we have a vector of rows.
- We just return `&self.periodic_columns[self.row_index % col.len()]`